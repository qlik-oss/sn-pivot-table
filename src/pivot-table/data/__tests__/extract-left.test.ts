import NxDimCellType from '../../../types/QIX';
import extractLeft from '../extract-left';
import createNodes from './test-helper';

describe('extractLeft', () => {
  test('should handle empty qLeft array', () => {
    const qLeft: EngineAPI.INxPivotDimensionCell[] = [];

    const left = extractLeft(qLeft);

    expect(left).toHaveLength(0);
  });

  test('should extract left data with no nodes expanded', () => {
    const rowCount = 3;
    const qLeft = createNodes(rowCount, NxDimCellType.NX_DIM_CELL_NORMAL);
    const left = extractLeft(qLeft);

    expect(left).toMatchSnapshot();
  });

  test('should extract left data with first node expanded', () => {
    const rowCount = 3;
    const subNodesCount = 2;
    const qLeft = createNodes(rowCount, NxDimCellType.NX_DIM_CELL_NORMAL);
    const subNodes = createNodes(subNodesCount, NxDimCellType.NX_DIM_CELL_NORMAL);
    qLeft[0].qSubNodes = subNodes;
    qLeft[0].qCanCollapse = true;
    const left = extractLeft(qLeft);

    expect(left).toMatchSnapshot();
  });

  test('should extract left data when data tree has a depth of 2', () => {
    const rowCount = 3;
    const subNodesCount = 2;
    const qLeft = createNodes(rowCount, NxDimCellType.NX_DIM_CELL_NORMAL);

    qLeft[0].qSubNodes = createNodes(1, NxDimCellType.NX_DIM_CELL_EMPTY);
    qLeft[0].qSubNodes[0].qSubNodes = createNodes(1, NxDimCellType.NX_DIM_CELL_EMPTY);

    qLeft[1].qSubNodes = createNodes(1, NxDimCellType.NX_DIM_CELL_EMPTY);
    qLeft[1].qSubNodes[0].qSubNodes = createNodes(1, NxDimCellType.NX_DIM_CELL_EMPTY);

    qLeft[2].qSubNodes = createNodes(subNodesCount, NxDimCellType.NX_DIM_CELL_NORMAL);
    qLeft[2].qSubNodes[0].qSubNodes = createNodes(1, NxDimCellType.NX_DIM_CELL_EMPTY);
    qLeft[2].qCanCollapse = true;
    qLeft[2].qSubNodes[1].qSubNodes = createNodes(subNodesCount, NxDimCellType.NX_DIM_CELL_NORMAL);;
    qLeft[2].qSubNodes[1].qCanCollapse = true;

    const left = extractLeft(qLeft);

    expect(left).toMatchSnapshot();
  });
});
