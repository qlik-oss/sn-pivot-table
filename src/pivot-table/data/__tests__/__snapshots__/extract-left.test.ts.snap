// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractLeft should extract left data when data tree has a depth of 2 1`] = `
Array [
  Array [
    Object {
      "parent": null,
      "qElemNo": 0,
      "qSubNodes": Array [
        Object {
          "qElemNo": 0,
          "qSubNodes": Array [
            Object {
              "qElemNo": 0,
              "qSubNodes": Array [],
              "qType": "E",
            },
          ],
          "qType": "E",
        },
      ],
      "qType": "N",
      "x": 0,
      "y": 0,
    },
    Object {
      "parent": null,
      "qElemNo": 1,
      "qSubNodes": Array [
        Object {
          "qElemNo": 0,
          "qSubNodes": Array [
            Object {
              "qElemNo": 0,
              "qSubNodes": Array [],
              "qType": "E",
            },
          ],
          "qType": "E",
        },
      ],
      "qType": "N",
      "x": 0,
      "y": 1,
    },
    Object {
      "parent": null,
      "qCanCollapse": true,
      "qElemNo": 2,
      "qSubNodes": Array [
        Object {
          "qElemNo": 0,
          "qSubNodes": Array [
            Object {
              "qElemNo": 0,
              "qSubNodes": Array [],
              "qType": "E",
            },
          ],
          "qType": "N",
        },
        Object {
          "qCanCollapse": true,
          "qElemNo": 1,
          "qSubNodes": Array [
            Object {
              "qElemNo": 0,
              "qSubNodes": Array [],
              "qType": "N",
            },
            Object {
              "qElemNo": 1,
              "qSubNodes": Array [],
              "qType": "N",
            },
          ],
          "qType": "N",
        },
      ],
      "qType": "N",
      "x": 0,
      "y": 2,
    },
  ],
  Array [
    Object {
      "parent": Object {
        "parent": null,
        "qElemNo": 0,
        "qSubNodes": Array [
          Object {
            "qElemNo": 0,
            "qSubNodes": Array [
              Object {
                "qElemNo": 0,
                "qSubNodes": Array [],
                "qType": "E",
              },
            ],
            "qType": "E",
          },
        ],
        "qType": "N",
        "x": 0,
        "y": 0,
      },
      "qElemNo": 0,
      "qSubNodes": Array [
        Object {
          "qElemNo": 0,
          "qSubNodes": Array [],
          "qType": "E",
        },
      ],
      "qType": "E",
      "x": 1,
      "y": 0,
    },
    Object {
      "parent": Object {
        "parent": null,
        "qElemNo": 1,
        "qSubNodes": Array [
          Object {
            "qElemNo": 0,
            "qSubNodes": Array [
              Object {
                "qElemNo": 0,
                "qSubNodes": Array [],
                "qType": "E",
              },
            ],
            "qType": "E",
          },
        ],
        "qType": "N",
        "x": 0,
        "y": 1,
      },
      "qElemNo": 0,
      "qSubNodes": Array [
        Object {
          "qElemNo": 0,
          "qSubNodes": Array [],
          "qType": "E",
        },
      ],
      "qType": "E",
      "x": 1,
      "y": 1,
    },
    Object {
      "parent": Object {
        "parent": null,
        "qCanCollapse": true,
        "qElemNo": 2,
        "qSubNodes": Array [
          Object {
            "qElemNo": 0,
            "qSubNodes": Array [
              Object {
                "qElemNo": 0,
                "qSubNodes": Array [],
                "qType": "E",
              },
            ],
            "qType": "N",
          },
          Object {
            "qCanCollapse": true,
            "qElemNo": 1,
            "qSubNodes": Array [
              Object {
                "qElemNo": 0,
                "qSubNodes": Array [],
                "qType": "N",
              },
              Object {
                "qElemNo": 1,
                "qSubNodes": Array [],
                "qType": "N",
              },
            ],
            "qType": "N",
          },
        ],
        "qType": "N",
        "x": 0,
        "y": 2,
      },
      "qElemNo": 0,
      "qSubNodes": Array [
        Object {
          "qElemNo": 0,
          "qSubNodes": Array [],
          "qType": "E",
        },
      ],
      "qType": "N",
      "x": 1,
      "y": 2,
    },
    Object {
      "parent": Object {
        "parent": null,
        "qCanCollapse": true,
        "qElemNo": 2,
        "qSubNodes": Array [
          Object {
            "qElemNo": 0,
            "qSubNodes": Array [
              Object {
                "qElemNo": 0,
                "qSubNodes": Array [],
                "qType": "E",
              },
            ],
            "qType": "N",
          },
          Object {
            "qCanCollapse": true,
            "qElemNo": 1,
            "qSubNodes": Array [
              Object {
                "qElemNo": 0,
                "qSubNodes": Array [],
                "qType": "N",
              },
              Object {
                "qElemNo": 1,
                "qSubNodes": Array [],
                "qType": "N",
              },
            ],
            "qType": "N",
          },
        ],
        "qType": "N",
        "x": 0,
        "y": 2,
      },
      "qCanCollapse": true,
      "qElemNo": 1,
      "qSubNodes": Array [
        Object {
          "qElemNo": 0,
          "qSubNodes": Array [],
          "qType": "N",
        },
        Object {
          "qElemNo": 1,
          "qSubNodes": Array [],
          "qType": "N",
        },
      ],
      "qType": "N",
      "x": 1,
      "y": 3,
    },
  ],
  Array [
    Object {
      "parent": Object {
        "parent": Object {
          "parent": null,
          "qElemNo": 0,
          "qSubNodes": Array [
            Object {
              "qElemNo": 0,
              "qSubNodes": Array [
                Object {
                  "qElemNo": 0,
                  "qSubNodes": Array [],
                  "qType": "E",
                },
              ],
              "qType": "E",
            },
          ],
          "qType": "N",
          "x": 0,
          "y": 0,
        },
        "qElemNo": 0,
        "qSubNodes": Array [
          Object {
            "qElemNo": 0,
            "qSubNodes": Array [],
            "qType": "E",
          },
        ],
        "qType": "E",
        "x": 1,
        "y": 0,
      },
      "qElemNo": 0,
      "qSubNodes": Array [],
      "qType": "E",
      "x": 2,
      "y": 0,
    },
    Object {
      "parent": Object {
        "parent": Object {
          "parent": null,
          "qElemNo": 1,
          "qSubNodes": Array [
            Object {
              "qElemNo": 0,
              "qSubNodes": Array [
                Object {
                  "qElemNo": 0,
                  "qSubNodes": Array [],
                  "qType": "E",
                },
              ],
              "qType": "E",
            },
          ],
          "qType": "N",
          "x": 0,
          "y": 1,
        },
        "qElemNo": 0,
        "qSubNodes": Array [
          Object {
            "qElemNo": 0,
            "qSubNodes": Array [],
            "qType": "E",
          },
        ],
        "qType": "E",
        "x": 1,
        "y": 1,
      },
      "qElemNo": 0,
      "qSubNodes": Array [],
      "qType": "E",
      "x": 2,
      "y": 1,
    },
    Object {
      "parent": Object {
        "parent": Object {
          "parent": null,
          "qCanCollapse": true,
          "qElemNo": 2,
          "qSubNodes": Array [
            Object {
              "qElemNo": 0,
              "qSubNodes": Array [
                Object {
                  "qElemNo": 0,
                  "qSubNodes": Array [],
                  "qType": "E",
                },
              ],
              "qType": "N",
            },
            Object {
              "qCanCollapse": true,
              "qElemNo": 1,
              "qSubNodes": Array [
                Object {
                  "qElemNo": 0,
                  "qSubNodes": Array [],
                  "qType": "N",
                },
                Object {
                  "qElemNo": 1,
                  "qSubNodes": Array [],
                  "qType": "N",
                },
              ],
              "qType": "N",
            },
          ],
          "qType": "N",
          "x": 0,
          "y": 2,
        },
        "qElemNo": 0,
        "qSubNodes": Array [
          Object {
            "qElemNo": 0,
            "qSubNodes": Array [],
            "qType": "E",
          },
        ],
        "qType": "N",
        "x": 1,
        "y": 2,
      },
      "qElemNo": 0,
      "qSubNodes": Array [],
      "qType": "E",
      "x": 2,
      "y": 2,
    },
    Object {
      "parent": Object {
        "parent": Object {
          "parent": null,
          "qCanCollapse": true,
          "qElemNo": 2,
          "qSubNodes": Array [
            Object {
              "qElemNo": 0,
              "qSubNodes": Array [
                Object {
                  "qElemNo": 0,
                  "qSubNodes": Array [],
                  "qType": "E",
                },
              ],
              "qType": "N",
            },
            Object {
              "qCanCollapse": true,
              "qElemNo": 1,
              "qSubNodes": Array [
                Object {
                  "qElemNo": 0,
                  "qSubNodes": Array [],
                  "qType": "N",
                },
                Object {
                  "qElemNo": 1,
                  "qSubNodes": Array [],
                  "qType": "N",
                },
              ],
              "qType": "N",
            },
          ],
          "qType": "N",
          "x": 0,
          "y": 2,
        },
        "qCanCollapse": true,
        "qElemNo": 1,
        "qSubNodes": Array [
          Object {
            "qElemNo": 0,
            "qSubNodes": Array [],
            "qType": "N",
          },
          Object {
            "qElemNo": 1,
            "qSubNodes": Array [],
            "qType": "N",
          },
        ],
        "qType": "N",
        "x": 1,
        "y": 3,
      },
      "qElemNo": 0,
      "qSubNodes": Array [],
      "qType": "N",
      "x": 2,
      "y": 3,
    },
    Object {
      "parent": Object {
        "parent": Object {
          "parent": null,
          "qCanCollapse": true,
          "qElemNo": 2,
          "qSubNodes": Array [
            Object {
              "qElemNo": 0,
              "qSubNodes": Array [
                Object {
                  "qElemNo": 0,
                  "qSubNodes": Array [],
                  "qType": "E",
                },
              ],
              "qType": "N",
            },
            Object {
              "qCanCollapse": true,
              "qElemNo": 1,
              "qSubNodes": Array [
                Object {
                  "qElemNo": 0,
                  "qSubNodes": Array [],
                  "qType": "N",
                },
                Object {
                  "qElemNo": 1,
                  "qSubNodes": Array [],
                  "qType": "N",
                },
              ],
              "qType": "N",
            },
          ],
          "qType": "N",
          "x": 0,
          "y": 2,
        },
        "qCanCollapse": true,
        "qElemNo": 1,
        "qSubNodes": Array [
          Object {
            "qElemNo": 0,
            "qSubNodes": Array [],
            "qType": "N",
          },
          Object {
            "qElemNo": 1,
            "qSubNodes": Array [],
            "qType": "N",
          },
        ],
        "qType": "N",
        "x": 1,
        "y": 3,
      },
      "qElemNo": 1,
      "qSubNodes": Array [],
      "qType": "N",
      "x": 2,
      "y": 4,
    },
  ],
]
`;

exports[`extractLeft should extract left data with first node expanded 1`] = `
Array [
  Array [
    Object {
      "parent": null,
      "qCanCollapse": true,
      "qElemNo": 0,
      "qSubNodes": Array [
        Object {
          "qElemNo": 0,
          "qSubNodes": Array [],
          "qType": "N",
        },
        Object {
          "qElemNo": 1,
          "qSubNodes": Array [],
          "qType": "N",
        },
      ],
      "qType": "N",
      "x": 0,
      "y": 0,
    },
    Object {
      "parent": null,
      "qElemNo": 1,
      "qSubNodes": Array [],
      "qType": "N",
      "x": 0,
      "y": 2,
    },
    Object {
      "parent": null,
      "qElemNo": 2,
      "qSubNodes": Array [],
      "qType": "N",
      "x": 0,
      "y": 3,
    },
  ],
  Array [
    Object {
      "parent": Object {
        "parent": null,
        "qCanCollapse": true,
        "qElemNo": 0,
        "qSubNodes": Array [
          Object {
            "qElemNo": 0,
            "qSubNodes": Array [],
            "qType": "N",
          },
          Object {
            "qElemNo": 1,
            "qSubNodes": Array [],
            "qType": "N",
          },
        ],
        "qType": "N",
        "x": 0,
        "y": 0,
      },
      "qElemNo": 0,
      "qSubNodes": Array [],
      "qType": "N",
      "x": 1,
      "y": 0,
    },
    Object {
      "parent": Object {
        "parent": null,
        "qCanCollapse": true,
        "qElemNo": 0,
        "qSubNodes": Array [
          Object {
            "qElemNo": 0,
            "qSubNodes": Array [],
            "qType": "N",
          },
          Object {
            "qElemNo": 1,
            "qSubNodes": Array [],
            "qType": "N",
          },
        ],
        "qType": "N",
        "x": 0,
        "y": 0,
      },
      "qElemNo": 1,
      "qSubNodes": Array [],
      "qType": "N",
      "x": 1,
      "y": 1,
    },
  ],
]
`;

exports[`extractLeft should extract left data with no nodes expanded 1`] = `
Array [
  Array [
    Object {
      "parent": null,
      "qElemNo": 0,
      "qSubNodes": Array [],
      "qType": "N",
      "x": 0,
      "y": 0,
    },
    Object {
      "parent": null,
      "qElemNo": 1,
      "qSubNodes": Array [],
      "qType": "N",
      "x": 0,
      "y": 1,
    },
    Object {
      "parent": null,
      "qElemNo": 2,
      "qSubNodes": Array [],
      "qType": "N",
      "x": 0,
      "y": 2,
    },
  ],
]
`;
